test/contrib/automotive/interface_mockup.py

```diff
@@ -209,3 +222,79 @@ def exit_if_no_isotp_module():
 from scapy.contrib.automotive.ecu import *
 log_runtime.debug("Set send delay to lower utilization on CI machines")
 conf.contribs['EcuAnsweringMachine']['send_delay'] = 0.004
+
+# ############################################################################
+# """ Define custom SuperSocket for unit tests """
+# ############################################################################
+
+
+class TestSocket(ObjectPipe, object):
+    nonblocking_socket = True  # type: bool
+
+    def __init__(self, basecls=None):
+        # type: (Optional[Type[Packet]]) -> None
+        super(TestSocket, self).__init__()
+        self.basecls = basecls
+        self.__paired_socket = None  # type: Optional[TestSocket]
+        self.closed = False
+
+    def close(self):
+        self.closed = True
+        super(TestSocket, self).close()
+
+    def pair(self, sock):
+        # type: (TestSocket) -> None
+        if sock.__paired_socket or self.__paired_socket:
+            raise Scapy_Exception("Socket already paired")
+        self.__paired_socket = sock
+        sock.__paired_socket = self
+
+    def send(self, x):
+        # type: (Packet) -> int
+        if not self.__paired_socket:
+            self.close()
+            raise Scapy_Exception("Socket not paired!")
+
+        sx = bytes(x)
+        super(TestSocket, self.__paired_socket).send(sx)
+        try:
+            x.sent_time = time.time()
+        except AttributeError:
+            pass
+        return len(sx)
+
+    def recv_raw(self, x=MTU):
+        # type: (int) -> Tuple[Optional[Type[Packet]], Optional[bytes], Optional[float]]  # noqa: E501
+        """Returns a tuple containing (cls, pkt_data, time)"""
+        return self.basecls, \
+            super(TestSocket, self).recv(), \
+            time.time()
+
+    def recv(self, x=MTU):
+        # type: (int) -> Optional[Packet]
+        cls, val, ts = self.recv_raw(x)
+        if not val or not cls:
+            return None
+        try:
+            pkt = cls(val)  # type: Packet
+        except KeyboardInterrupt:
+            raise
+        except Exception:
+            if conf.debug_dissector:
+                from scapy.sendrecv import debug
+                debug.crashed_on = (cls, val)
+                raise
+            pkt = conf.raw_layer(val)
+        if ts:
+            pkt.time = ts
+        return pkt
```

Comment:
Have I not already reviewed this? This function is a duplicate

Reply:
applied in #3309